### DISCUSSION TITLE
Linking external libraries across different modules

### DISCUSSION THREAD

**Discussion by mcacace** [idaholab/moose#18834] [Q&A General] [answered]
Dear all,
I linked an external library (mfront) to the tensor mechanics module. All worked fine and my tests did run with success. Now, I would like to combine this module with others (e. g. porous flow and navier stokes). Would it be enough to activate the module or should I add all relevant flags somewhere specific. I am saying this since a first attempt to make porous flow threw a linking error while making the tensor mechanics module (all external calls are no longer valid). Need to say that I added all flags to the make file of tensor mechanics. 
Thanks for any feedback,
Mauro

**Comment by GiudGiud** [comment]
Should be enough to activate the module in your makefile for your applications. 
If you are running with the modules directly, you may want to use the combined module

**Reply by GiudGiud** [reply]
linking errors are common if you havent cleaned your repository of pre-compiled objects before re-compiling with different sets of modules

**Reply by mcacace** [reply]
Sorry to reopen this post (BTW, thanks @GiudGiud for the reply), but even after retrying - force cleaning and remake the combined module, I am still facing the same problem. 
If I compile the tensor mechanics module directly all works fine, however if I try to compile it via the combined module I keep receiving errors for undefined references to specific calls of the external library I want to link. Not sure, what is wrong.
As always, any help would be appreciated. 
Thanks,
Mauro 

For the sake of completeness, these are the additional flags I required (added to the makefile):

ADDITIONAL_CPPFLAGS += -I/home/cacace/snippets/MFrontGenericInterfaceSupport/include/
ADDITIONAL_LIBS += -L/home/cacace/snippets/MFrontGenericInterfaceSupport/build/src/ -lMFrontGenericInterface

**Reply by GiudGiud** [reply]
can you please paste the error message?
Did you `make clobberall` before compiling the combined module?

**Reply by mcacace** [reply]
I did force make clobberall before recompiling, but it did not work. 
Attached the screenshot of the error.
![error](https://user-images.githubusercontent.com/13695024/136818976-8efba8c9-b5c3-49d6-9aae-6cce82b8abe3.png)

**Reply by mcacace** [reply]
All function related to the mgis library are undefined...

**Reply by GiudGiud** [reply]
can you save all your work (git commit) then run git clean -xfd in the repo

**Reply by mcacace** [reply]
@GiudGiud Thanks! I will try and let you know.

**Reply by GiudGiud** [reply]
this flag is a little weird btw
ADDITIONAL_LIBS += -L/home/cacace/snippets/MFrontGenericInterfaceSupport/build/src/
I d expect a lib/ folder not a src/folder

**Reply by anilkunwar** [reply]
@mcacace  I am curious to know whether you can access the individual modules from external user app. When did you create your external app if you have created one . I have encountered some issues with the newly created (created few months ago) external app. I do not see the necessity to compile the individual modules as long as the user app can access them externally through compilation.  Previously it was recommended to  use the modules through an external app outside the framework, and so I had never compiled the individual modules themselves. Recently,  when  an older external app (created before 2020) compiled on top of recently installed MOOSE  could not  access all of the modules  of MOOSE, I first tried to compile the individual modules by jumping into the modules directory inside the moose directory and I have never been successful with the combined module. so I choose to compile ( individualmodulename := yes in the Makefile) of the external app and it has been successful in accessing phase field, navier stokes, heat conduction, tensor mechanics and xfem modules. 
This explanation is associated with moose installation in Ubuntu 20.04 in conda environment.  `ALL_MODULES := yes `is still working fine in the same older app compiled on top of MOOSE framework  installed in the previous Moose environment based installation in Ubuntu 18.04 OS.

**Comment by anilkunwar** [comment]
when I use `ALL_MODULES :=yes`, I also get the linking error for the POROUS_FLOW. For now I am using ALL_MODULES :=no and specifying the other linkable modules only 
```
PHASE_FIELD :=yes
NAVIER_STOKES :=yes
...
POROUS_FLOW :=no
...

```
In my case, even after the test runs successful with the executable of app, i cannot run the input file that uses objects from these compiled modules.

**Comment by dschwen** [comment, marked as answer]
@mcacace It seems to me a better approach would be to create your own app, and link the new library there. Your change to tensormechanics is unlikely to ever get merged into moose, and you'll be stuck rebasing your branch for all eternity. In https://github.com/idaholab/magpie we link a few external libraries successfully.

**Reply by mcacace** [reply]
@dschwen I also think that creating a simple app could solve the issue - indeed I already have the mfront library linked to another app we developed. I will discuss with my colleague, who is interested to make use of tensormechanics and porous flow, to see whether he is fine with it. BTW, the mfront library is an interesting one, and it does more or less the same as per the umat interface in the module - maybe it would be worth considering as an alternative.

**Reply by dschwen** [reply]
It does look interesting, however it is licensed under the GPL, which is incompatible with MOOSE's LGPL license. :-(
